### PR TITLE
Test database

### PR DESCRIPTION
**Pull request by dubhcait**
There weren't any issues unassigned so I gave this a go. 
I figured out how to make a test database the same as our heroku salesforce one. 
I had both DB query files on my local branch but didn't commit them as they are on other pull requests.

A lot of the learning is in setUpDB.js 

To use: 
Run `npm run buildtest` this creates a local database, schema, user and password
add TEST_DATABASE_URL= postgres://salesforcetester:test@localhost:5432/salesforcetest to your .env 
Run `npm test`

### CODE REVIEW

**Comment by dubhcait** [comment]
will need modification when the master is up to date after the new PRs are merged

File: server/db/testDB/build.sql [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+CREATE TABLE salesforce.application__c (
+id SERIAL PRIMARY KEY,
+question_1__c TEXT NOT NULL,
+question_2__c TEXT NOT NULL, 
+question_3__c TEXT NOT NULL
+
+);
+
+CREATE TABLE salesforce.Question__c (
```
Comment (dubhcait) [inline server/db/testDB/build.sql]: to lowercase Q

**Comment by dubhcait** [comment]
change answer test re .row[0]

**Comment by dubhcait** [comment]
really interesting you cant have an ssl connection to a local database

File: server/test/routes/questions.js [hunk 2]
```diff
@@ -13,9 +13,9 @@ test('GET /api/v1/questions', function (assert) {
       .end(function (err, res) {
 
         const expectedLength = res.body.data.length;
-
+       
         assert.error(err, 'No error')
-        assert.equal(3, expectedLength, '/api/v1/question works fine!')
+        assert.equal(1, expectedLength, '/api/v1/question works fine!')
```
Comment (dubhcait) [inline server/test/routes/questions.js]: running on the test database - it has only one entry for the question table

File: server/db/config/connection.js [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
+  process.env.NODE_ENV === "test" ?
+  process.env.TEST_DATABASE_URL :
+  process.env.DATABASE_URL;
 
 if (!connectionString) throw new Error('DATABASE_URL is not found!');
 
 const options = {
   connectionString,
-  ssl: true,
+  ssl: !connectionString.includes("localhost")
```
Comment (AbdallahAmmar96) [inline server/db/config/connection.js]: you can do this 
```suggestion
  ssl: true
```

File: server/db/testDB/build.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+const dbConnection = require("../config/connection");
+const fs = require("fs");
```
Comment (AbdallahAmmar96) [inline server/db/testDB/build.js]: we require the core modules before your local modules 
```suggestion
const fs = require("fs");
const dbConnection = require("../config/connection");
```

File: server/db/testDB/build.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    // eslint-disable-next-line no-unused-vars
+    dbConnection.query(sql, (err, res) => {
+      if (err) reject(err);
+      // eslint-disable-next-line no-console
+      console.log("database being built");
+      resolve(true);
+    });
+  });
+  //runDbBuild();
+module.exports = runDbBuild;
```
Comment (AbdallahAmmar96) [inline server/db/testDB/build.js]: add new line

File: server/db/testDB/build.sql [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+);
+
+INSERT INTO salesforce.application__c (question_1__c,  question_2__c  ,question_3__c ) VALUES 
+('Test Answer 1', 'Test Answer 2', 'Test Answer 3');
+
+INSERT INTO salesforce.questions__c ( title__c, wordsLimit__c, tips__c, expectations__c) VALUES 
+('Test Question', '250', 'Test tips', 'Test expectations' );
+
+
+COMMIT;
```
Comment (AbdallahAmmar96) [inline server/db/testDB/build.sql]: add new line

File: server/db/testDB/build.sql [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+CREATE TABLE salesforce.questions__c (
+id SERIAL PRIMARY KEY,
+title__c TEXT NOT NULL, 
+wordsLimit__c INT NOT NULL, 
+tips__c TEXT, 
+expectations__c TEXT
+
+);
+
+INSERT INTO salesforce.application__c (question_1__c,  question_2__c  ,question_3__c ) VALUES
```
Comment (AbdallahAmmar96) [inline server/db/testDB/build.sql]: you can split this code in another file one for the tables and the other for the fake data you have

File: server/db/testDB/setUpDB.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      console.log(error);
+    } else {
+      console.log("Created database salesforcetest");
+    }
+    runDbBuild()
+      .then(() => {
+        console.log("Successfully populated test database");
+        process.exit(0); 
+      })
+      .catch(console.log(error));
```
Comment (AbdallahAmmar96) [inline server/db/testDB/setUpDB.js]:
```suggestion
      .catch(console.log);
```

File: server/db/config/connection.js (hunk 3 above)
Comment (dubhcait) [reply server/db/config/connection.js]: afraid I can't when I set it to true I kept getting this when trying to run the tests


Local databases dont always support SSL

File: server/db/testDB/setUpDB.js (hunk 8 above)
Comment (dubhcait) [reply server/db/testDB/setUpDB.js]: interesting I haven't seen it used like that before

File: server/db/testDB/build.sql (hunk 7 above)
Comment (dubhcait) [reply server/db/testDB/build.sql]: I'm having issues with the running order when I'm splitting it up. I think I've to create it as two promises and resolve it with a promise.all(), is there a way I'm missing?

**Pull request merged**